File: apps/dimi/src/tui/components/messages/status-message.ts
import { Container, Spacer, Text } from '@dimi-agent/pi-tui';

import { currentTheme } from '#/tui/theme';
import type { ColorToken } from '#/tui/theme';

export class StatusMessageComponent extends Container {
  private textComponent: Text;
  private content: string;
  private color?: ColorToken;

  constructor(content: string, color?: ColorToken) {
    super();
    this.content = content;
    this.color = color;
    this.textComponent = new Text(this.renderText(), 0, 0);
    this.addChild(this.textComponent);
  }

  // Update the body in place (used for live-streamed `!` shell output) without
  // remounting the component.
  updateContent(content: string): void {
    this.content = content;
    this.textComponent.setText(this.renderText());
  }

  override invalidate(): void {
    this.textComponent.setText(this.renderText());
    super.invalidate();
  }

  // Indent every line, not just the first. The `content` may be multi-line
  // (e.g. `!` shell output); prefixing the whole string once would only indent
  // the first line and leave the rest at column 0. Strip carriage returns
  // first: a trailing `\r` (e.g. from CRLF server error pages) is zero-width
  // for the line wrapper, so the padding spaces appended after it overwrite
  // the visible content and the line renders blank.
  private renderText(): string {
    const colored =
      this.color === undefined
        ? currentTheme.fg('textDim', this.content)
        : currentTheme.fg(this.color, this.content);
    return colored.replaceAll('\r', '').split('\n').map((line) => `  ${line}`).join('\n');
  }
}

export class NoticeMessageComponent extends Container {
  private titleText: Text;
  private detailText?: Text;
  private title: string;
  private detail?: string;

  constructor(title: string, detail: string | undefined) {
    super();
    this.title = title;
    this.detail = detail;
    this.addChild(new Spacer(1));
    this.titleText = new Text(`  ${currentTheme.fg('textStrong', title)}`, 0, 0);
    this.addChild(this.titleText);
    if (detail !== undefined && detail.length > 0) {
      this.detailText = new Text(`  ${currentTheme.fg('textDim', detail)}`, 0, 0);
      this.addChild(this.detailText);
    }
  }

  override invalidate(): void {
    this.titleText.setText(`  ${currentTheme.fg('textStrong', this.title)}`);
    if (this.detailText !== undefined && this.detail !== undefined) {
      this.detailText.setText(`  ${currentTheme.fg('textDim', this.detail)}`);
    }
    super.invalidate();
  }
}

File: crates/dimi-tui/src/components/messages/status_message.rs
//! `StatusMessageComponent` / `NoticeMessageComponent` — status and notice
//! rows (port of `apps/dimi/src/tui/components/messages/status-message.ts`).

use crate::component::Component;
use crate::components::messages::MESSAGE_INDENT;
use crate::components::spacer::Spacer;
use crate::components::text::Text;
use crate::theme::{ColorToken, current_theme};

/// A single status line: indented, dimmed (or error/warning colored) text.
pub struct StatusMessageComponent {
    content: String,
    color: Option<ColorToken>,
    text_component: Text,
}

impl StatusMessageComponent {
    pub fn new(content: &str, color: Option<ColorToken>) -> Self {
        let text = render_text(content, color);
        StatusMessageComponent {
            content: content.to_owned(),
            color,
            text_component: Text::new(&text, 0, 0),
        }
    }

    /// Update the body in place (used for live-streamed `!` shell output).
    pub fn update_content(&mut self, content: &str) {
        self.content = content.to_owned();
        let text = render_text(content, self.color);
        self.text_component.set_text(&text);
    }
}

/// Indent every line (not just the first) and strip carriage returns.
fn render_text(content: &str, color: Option<ColorToken>) -> String {
    let colored = match color {
        Some(token) => current_theme().fg(token, content),
        None => current_theme().fg(ColorToken::TextDim, content),
    };
    colored
        .replace('\r', "")
        .split('\n')
        .map(|line| format!("{MESSAGE_INDENT}{line}"))
        .collect::<Vec<_>>()
        .join("\n")
}

impl Component for StatusMessageComponent {
    fn render(&mut self, width: usize) -> Vec<String> {
        self.text_component.render(width)
    }

    fn invalidate(&mut self) {
        let text = render_text(&self.content, self.color);
        self.text_component.set_text(&text);
    }
}

/// A notice: blank line + title + optional detail.
pub struct NoticeMessageComponent {
    title: String,
    detail: Option<String>,
    title_text: Text,
    detail_text: Option<Text>,
    has_spacer: bool,
}

impl NoticeMessageComponent {
    pub fn new(title: &str, detail: Option<&str>) -> Self {
        let title_text = Text::new(
            &format!(
                "{MESSAGE_INDENT}{}",
                current_theme().fg(ColorToken::TextStrong, title)
            ),
            0,
            0,
        );
        let detail_text = detail.filter(|d| !d.is_empty()).map(|d| {
            Text::new(
                &format!(
                    "{MESSAGE_INDENT}{}",
                    current_theme().fg(ColorToken::TextDim, d)
                ),
                0,
                0,
            )
        });
        NoticeMessageComponent {
            title: title.to_owned(),
            detail: detail.map(str::to_owned),
            title_text,
            detail_text,
            has_spacer: true,
        }
    }
}

impl Component for NoticeMessageComponent {
    fn render(&mut self, width: usize) -> Vec<String> {
        let mut lines = Vec::new();
        if self.has_spacer {
            lines.extend(Spacer::new(1).render(width));
        }
        lines.extend(self.title_text.render(width));
        if let Some(detail) = &mut self.detail_text {
            lines.extend(detail.render(width));
        }
        lines
    }

    fn invalidate(&mut self) {
        self.title_text.set_text(&format!(
            "{MESSAGE_INDENT}{}",
            current_theme().fg(ColorToken::TextStrong, &self.title)
        ));
        if let (Some(detail), Some(detail_text)) = (&self.detail, &mut self.detail_text) {
            detail_text.set_text(&format!(
                "{MESSAGE_INDENT}{}",
                current_theme().fg(ColorToken::TextDim, detail)
            ));
        }
    }
}
